### PR TITLE
feat/fix: exhaustive graph validation & session processor soft-lock

### LINKED ISSUE #4744: [bug]: Queue generation stalls but doesn't halt when fields are incompatible
### Is there an existing issue for this?

- [X] I have searched the existing issues

### OS

macOS

### GPU

mps

### VRAM

16

### What version did you experience this issue on?

main

### What happened?

When I set outputs on a node I was writing to `Optional[]` and connected one to the IP-Adapter weight input, and float primitive input, I got the following error in the terminal when invoking:
```
[2023-09-29 21:17:15,831]::[InvokeAI]::ERROR --> Error in session processor: Fields are incompatible: cannot connect 7827adc8-7371-4aa3-81e7-31f685fcc4ea.float_output to 5ea62fe4-06c4-4f45-8f97-a056a352b5ff.weight
```
Seems it's just connection validation complaining that something is marked as optional.

However, there were no JS console errors, the execution didn't halt (just stalled with queue progress bar in fast "thinking/processing" mode), and the queue progress remained as "In Progress". 

### Screenshots

workflow:
Screenshot_2023-09-29_at_9 17 31_PM
offending node (`valuetovalue.py`):
[valuetovalue.py.zip](https://github.com/invoke-ai/InvokeAI/files/12772195/valuetovalue.py.zip)

### Additional context

_No response_

### Contact Details

_No response_

### PR DESCRIPTION
**Pull request by psychedelicious**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [ ] Yes
- [x] No, because:

## Description

[feat(nodes): exhaustive graph validation](https://github.com/invoke-ai/InvokeAI/commit/d2a88a47bde7dde297b5fc4f328d28ddfdeea1a4)

Makes graph validation logic more rigorous, validating graphs when they are created as part of a session or batch.

`validate_self()` method added to `Graph` model. It does all the validation that `is_valid()` did, plus a few extras:
- unique `node.id` values across graph
- node ids match their key in `Graph.nodes`
- recursively validate subgraphs
- validate all edges
- validate graph is acyclical

The new method is required because `is_valid()` just returned a boolean. That behaviour is retained, but `validate_self()` now raises appropriate exceptions for validation errors. This are then surfaced to the client.

The function is named `validate_self()` because pydantic reserves `validate()`.

There are two main places where graphs are created - in batches and in sessions.

Field validators are added to each of these for their `graph` fields, which call the new validation logic.

**Closes #4744**

In this issue, a batch is enqueued with an invalid graph. The output field is typed as optional while the input field is required. The field types themselves are not relevant - this change addresses the case where an invalid graph was created.

The mismatched types problem is not noticed until we attempt to invoke the graph, because the graph was never *fully* validated. An error is raised during the call to `graph_execution_state.next()` in `invoker.py`. This function prepares the edges and validates them, raising an exception due to the mismatched types.

This exception is caught by the session processor, but it doesn't handle this situation well - the graph is not marked as having an error and the queue item status is never changed. The queue item is therefore forever `in_progress`, so no new queue items are popped - the app won't do anything until the queue item is canceled manually.

This commit addresses this by preventing invalid graphs from being created in the first place, addressing a substantial number of fail cases.

[fix(nodes): mark session queue items failed on processor error](https://github.com/invoke-ai/InvokeAI/commit/786ac7126b79626c76b614280865501efad3e6b6)

When the processor has an error and it has a queue item, mark that item failed.

This addresses processor errors resulting in `in_progress` queue items, which create a soft lock of the processor, requiring the user to cancel the `in_progress` item before anything else processes.

@maryhipp @brandonrising There are a couple small commercial changes required for this:
- Add an error arg to the session queue `cancel_queue_item` method
- Call this in the processor when there is an error on a queue item
- The error will be surfaced to UI, it should just be `str(e)`

## Related Tickets & Documents

- Related Issue #
- Closes #4744

### Testing the graph validation changes

Test graph (you will need to disable validation in the workflow editor settings to connect the edge)

image

On `main`, this will soft-lock the processor, resulting in a perpetually `in_progress` queue item. You'd need to cancel it manually for the app to continue processing the queue.

On this pr, no queue item should be created, and you should get a toast saying that the edge is invalid.

### Testing the processor soft-lock fix

I'm not sure how to create an error directly in the processor on this PR (which is a good thing).

To test soft-lock fix in this PR, check out `main` and cherry-pick only the second commit 786ac7126b79626c76b614280865501efad3e6b6.

Run the same graph from the previous test and and you should see the queue item fails immediately.

There is not UI feedback for this situation yet -  I need to think about how to handle it. Probably needs a new event. But a processor exception like this should now be exceedingly rare, thanks to the enhanced graph validation changes.

## Added/updated tests?

- [x] Yes
- [ ] No : _please replace this line with details on why tests
      have not been included_

### CODE REVIEW

File: invokeai/app/services/graph.py [hunk 1]
```diff
@@ -377,53 +336,108 @@ def delete_edge(self, edge: Edge) -> None:
...
+        - `NodeIdMismatchError`
+        - `InvalidSubGraphError`
+        - `NodeNotFoundError`
+        - `NodeFieldNotFoundError`
+        - `CyclicalGraphError`
+        - `InvalidEdgeError`
+        """
+
+        # Validate that all node ids are unique
+        node_ids = [n.id for n in self.nodes.values()]
```
Comment (brandonrising) [inline invokeai/app/services/graph.py:354]: if len(node_ids) > 1, should all nodes be on an edge?

Comment (psychedelicious) [reply invokeai/app/services/graph.py:354]: No, you can have a graph with N nodes and zero edges

Comment (psychedelicious) [reply invokeai/app/services/graph.py:354]: It's allowed - ex a single resize image node

Comment (psychedelicious) [reply invokeai/app/services/graph.py:354]: note that all of this validation logic already existed, it's just been pulled into a single function now and called when the session/batch is created. every session within the batch is also validated. should be fairly difficult to get invalid graphs in

**Review by hipsterusername — APPROVED** [review]

File: invokeai/app/services/graph.py (hunk 1 above)
Comment (brandonrising) [reply invokeai/app/services/graph.py:354]: Weird I looked at this today and didn't see any comments but I can see them on my phone

**Review by brandonrising — APPROVED** [review]

**Pull request merged**